test/integration/publish_int_test.go

```diff
@@ -0,0 +1,252 @@
+					`email: author-email@domain.tld`,
+				},
+				0,
+			},
+		},
+		{
+			"New ingredient with meta file and flags",
+			input{
+				[]string{"{{.MetaFile}}", tempFile, "--name", "im-a-name-from-flag", "--author", "author-name-from-flag <author-email-from-flag@domain.tld>"},
+				p.StrP(`
+name: im-a-name
+namespace: {{.Username}}/shared
+version: 2.3.4
+description: im-a-description
+authors:
+  - name: author-name 
+    email: author-email@domain.tld
+`),
+				nil,
+				true,
+			},
+			expect{
+				[]string{
+					`name: im-a-name-from-flag`,
+					`namespace: {{.Username}}/shared`,
+					`version: 2.3.4`,
+					`description: im-a-description`,
+					`name: author-name-from-flag`,
+					`email: author-email-from-flag@domain.tld`,
+				},
+				0,
+			},
+		},
+		{
+			"New ingredient with editor flag",
+			input{
+				[]string{"", tempFile, "--editor"},
+				nil,
+				p.StrP(`
+name: im-a-name
+namespace: {{.Username}}/shared
+version: 2.3.4
+description: im-a-description
+authors:
+  - name: author-name 
+    email: author-email@domain.tld
+`),
+				true,
+			},
+			expect{
+				[]string{
+					`name: im-a-name`,
+					`namespace: {{.Username}}/shared`,
+					`version: 2.3.4`,
+					`description: im-a-description`,
+					`name: author-name`,
+					`email: author-email@domain.tld`,
+				},
+				0,
+			},
+		},
+		{
+			"Cancel upload",
+			input{
+				[]string{"", tempFile, "--name", "bogus", "--namespace", "{{.Username}}/shared"},
+				nil,
+				nil,
+				false,
+			},
+			expect{
+				[]string{`name: bogus`},
+				0,
+			},
+		},
+		// --edit tests are currently not addressed, tracked here: https://activestatef.atlassian.net/browse/DX-1944
+	}
+	for _, tt := range tests {
+		suite.Run(tt.name, func() {
+			ts := e2e.New(suite.T(), false)
+			defer ts.Close()
+
+			ts.Env = append(ts.Env,
+				// Publish tests shouldn't run against staging as they pollute the inventory db and artifact cache
+				constants.APIHostEnvVarName+"="+os.Getenv(constants.APIHostEnvVarName),
+			)
+
+			user := ts.CreateNewUser()
+
+			templateVars := map[string]interface{}{
+				"Username": user.Username,
+				"Email":    user.Email,
+			}
+
+			if tt.input.metafile != nil {
+				inputMetaParsed, err := strutils.ParseTemplate(*tt.input.metafile, templateVars, nil)
+				suite.Require().NoError(err)
+				metafile, err := fileutils.WriteTempFile("metafile.yaml", []byte(inputMetaParsed))
+				suite.Require().NoError(err)
+				templateVars["MetaFile"] = metafile
+			}
+
+			args := make([]string, len(tt.input.args))
+			copy(args, tt.input.args)
+
+			for k, v := range args {
+				vp, err := strutils.ParseTemplate(v, templateVars, nil)
+				suite.Require().NoError(err)
+				args[k] = vp
+			}
+
+			cp := ts.SpawnWithOpts(
+				e2e.WithArgs(append([]string{"publish"}, args...)...),
+			)
+
+			// Send custom input via --editor
+			if tt.input.editorValue != nil {
+				cp.Expect("Press enter when done editing")
+				rx := regexp.MustCompile(`file:\s*?(.*?)\.\s`)
```

Comment:
I believe convention is to put this in the top-level scope as a constant so we catch it at compile time, not runtime. However, this is in an integration test, so I'll leave it up to you.

Reply:
Yeah with integration tests I tend to take more liberties, but fair point! Will address.